### ISSUE TITLE
$ErrorActionPreference and $Error automatic variable

### ISSUE THREAD

**Issue by metablaster** [MicrosoftDocs/PowerShell-Docs#5452]
# Documentation missing information

## Documentation missing side effects of setting `$ErrorActionPreference` to "Stop"

These variables are mentioned here:

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7

> $Error
> Contains an array of error objects that represent the most recent errors. The most recent error is the first error object in the array $Error[0].

> $ErrorActionPreference
> ...
> The valid values are as follows:
> ...
> Stop: Displays the error message and stops executing.

- Issue affects the following content:

  **Cmdlet reference & about_ topics**
  - [x] Version 7.x preview content
  - [x] Version 7.0 content
  - [x] Version 6 content
  - [x] Version 5.1 content

Both of these descriptions fail to mention that setting the `$ErrorActionPreference` to **Stop** will bloat the `$Error` variable with an additional error element, which makes the docs both wrong and incomplete.

Example code:

```
function Test-Error
{
	[CmdletBinding()]
	param()
	Write-Error -Message "Error message" -Category PermissionDenied -ErrorId SampleID
}

$Error.Clear()
$ErrorActionPreference = "Stop"

Test-Error
```

According to docs this should produce exactly one error, that is the `$Error[0]`

However the automatic variable will be spammed with 2 errors of because preference is "Stop":

`$Error[0]` and `$Error[1]`

Basically we have 2 errors generated for single `Write-Error`, other error action preferences seem to produce only one error element as the docs are explaining.

**Comment by theJasonHelmick** [comment]
@metablaster Thanks for pointing this out and I will correct in the documentation. It seems that PowerShell throws an error (as expected) due to the $errorActionPreference = 'Stop' along with the actual error.  This is expected behavior.

**Comment by theJasonHelmick** [comment]
@sdwheeler 
When an error occurs in an interactive session or script, the error record details are stored in the array variable `$Error` and may be viewed using the `Get-Error` cmdlet. If you set the preference variable `$ErrorActionPreference ='Stop'`, PowerShell generates an additional error based on the preference variable. Both error record details can be viewed with `Get-Error -Newest 2`

**Comment by metablaster** [comment]
Thank you for feedback, I really didn't know that.

However one problem I see is, what happens when you want to get last 3 or 4 errors?

there is no way to tell if some of these 3 or 4 are having entries for "Stop" preference too, because the preference can be placed anywhere in the code.


but that should be probably asked somewhere else.

**Comment by sdwheeler** [comment]
@metablaster You can look at the type of the error record. Using your example, you see the following:

```
PS> $Error[0].GetType().Name
ActionPreferenceStopException
PS> $Error[1].GetType().Name
ErrorRecord
PS> $Error[0]

ErrorRecord                 : Error message
WasThrownFromThrowStatement : False
TargetSite                  : System.Collections.ObjectModel.Collection`1[System.Management.Automation.P
                              SObject] Invoke(System.Collections.IEnumerable)
StackTrace                  :    at
                              System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable
                              input)
                                 at Microsoft.PowerShell.Executor.ExecuteCommandHelper(Pipeline
                              tempPipeline, Exception& exceptionThrown, ExecutionOptions options)
Message                     : The running command stopped because the preference variable
                              "ErrorActionPreference" or common parameter is set to Stop: Error message
Data                        : {System.Management.Automation.Interpreter.InterpretedFrameInfo}
InnerException              :
HelpLink                    :
Source                      : System.Management.Automation
HResult                     : -2146233087


PS> $Error[0].ErrorRecord | fl -Force

Exception             : Microsoft.PowerShell.Commands.WriteErrorException: Error message
TargetObject          :
CategoryInfo          : PermissionDenied: (:) [Write-Error], WriteErrorException
FullyQualifiedErrorId : SampleID,Test-Error
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Test-Error, <No file>: line 5
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {0, 1}
PSMessageDetails      :


PS> $Error[1] | fl -Force

Exception             : Microsoft.PowerShell.Commands.WriteErrorException: Error message
TargetObject          :
CategoryInfo          : PermissionDenied: (:) [Write-Error], WriteErrorException
FullyQualifiedErrorId : SampleID,Test-Error
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Test-Error, <No file>: line 5
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}
PSMessageDetails      :
```

You can drill into the `ErrorRecord` or `InnerException` properties (as appropriate) of `$Error[0]` to get at the source error condition.

**Comment by metablaster** [comment]
```
PS> $Error[0].GetType().Name
ActionPreferenceStopException
PS> $Error[1].GetType().Name
ErrorRecord
```

@sdwheeler thank you for letting me know, just these 2 lines look like enough to loop trough `Error` variable and get the desired result.
